data_lang/pretty.py

```diff
@@ -1,36 +1,145 @@
 
 _ = log
 
+# TODO:
+# - options: max_depth, max_lines, stuff about cycles, LOSSY_JSON
+# - clean up imports (is there a lint that checks for unused imports?)
+# - hook up the printer in core/ui.py::PrettyPrintValue
+# - run the linter
+# - what's with `_ = log`?
 
-def FromValue(val):
-    # type: (value_t) -> doc_t
-    """Stub to turn an Oils value into a PPL.
+LOSSY_JSON = True
+
+class PrettyPrinter(object):
+    """Pretty print an Oils value.
+
+    Uses a strict version of the algorithm from Wadler's "A Prettier Printer".
+    (https://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf)
+    (https://lindig.github.io/papers/strictly-pretty-2000.pdf)
     """
-    with tagswitch(val) as case:
-        # e.g. see data_lang/j8.py
-        # It has options like SHOW_CYCLES and SHOW_NON_DATA
 
+    DEFAULT_MAX_WIDTH = 80
+
+    def __init__(self):
+        # type: () -> None
+        """Construct a PrettyPrinter with default configuration options.
+
+        Use the Init_*() methods for configuration before printing."""
+        self.max_width = PrettyPrinter.DEFAULT_MAX_WIDTH
+
+    def Init_MaxWidth(self, max_width):
+        # type: (int) -> None
+        self.max_width = max_width
+
+    def PrintValue(self, val, buf):
+        # type: (value_t, BufWriter) -> None
+        """Pretty print an Oils value to a BufWriter."""
+
+        document = _ValueToDoc(val)
+        self._PrintDoc(document, buf)
+
+    def _PrintDoc(self, document, buf):
+        # type: (doc_t, BufWriter) -> None
+        """Pretty print a `pretty.doc` to a BufWriter."""
+
+        fragments = [DocFragment(document, 0, False)]
+
+        while len(fragments) > 0:
+            frag = fragments.pop()
+            with tagswitch(frag.node) as case:
+
+                if case(doc_e.Newline):
+                    if frag.flat:
+                        buf.write(' ')
+                    else:
+                        buf.write('\n')
+                        buf.write_spaces(frag.indent)
+
+                elif case(doc_e.Text):
+                    text = cast(doc.Text, frag.node)
+                    buf.write(text.string)
+
+                elif case(doc_e.Indent):
+                    indented = cast(doc.Indent, frag.node)
+                    fragments.append(DocFragment(
+                        indented.node,
+                        frag.indent + indented.indent,
+                        frag.flat
+                    ))
+
+                elif case(doc_e.Concat):
+                    concat = cast(doc.Concat, frag.node)
+                    for node in reversed(concat.nodes):
+                        fragments.append(DocFragment(
+                            node,
+                            frag.indent,
+                            frag.flat
+                        ))
+
+def _ValueToDoc(val):
+    # type: (value_t) -> doc_t
+    """Convert an Oils value into a `doc`, which can then be pretty printed."""
+
+    with tagswitch(val) as case:
```

Comment:
I think we can fix the C++ compile errors using this pattern, basically add `UP_val = val` above this line, and then the downcasts are `val = cast(value.Bool, UP_val)` instead


https://www.oilshell.org/release/0.21.0/doc/oils-repo/mycpp/README.html#val-up_val-val-downcasting-pattern

Reply:
Oh, it's just about variable reuse! The all caps UP_ made me think there was something a lot more magic going on, like the compiler treating casts of things containing "UP_" differently.

I'll just give the casted values more specific names, as is appropriate since they have different types and hold different data.

Reply:
Actually mycpp DOES have a rule for `UP_` now :-P

```
$ grep 'UP_' mycpp/*.py
mycpp/cppgen_pass.py:            #    src = cast(source__SourcedFile, UP_src)
mycpp/cppgen_pass.py:            # -> source__SourcedFile* src = static_cast<source__SourcedFile>(UP_src)
mycpp/cppgen_pass.py:                    if to_cast.name.startswith('UP_'):
```

This is not really ideal -- there are threads on Zulip about "improving the cast pattern"

I'm open to work on that (remember I want to rewrite mycpp completely) -- but I guess SEPARATELY from this PR, since a lot of code would change

I think for this PR it's better to keep consistency with the `UP_val` style

It's possible some other schemes might work, I haven't thought about it

There is an assymmetry between vars "going in" to the switch and vars "coming out", and I think we mostly solved that, and it's documented